### PR TITLE
4 feature 3 user searchdiscovery

### PR DESCRIPTION
**Pull request by Christina0313**
The Search Discovery part is mostly completed. When clicking Search Discovery it will display the question generate from the use case and allow user to enter their answer. Then it will generate the list and present it. The presenting part need to connect with the other discovery cases, so not completed done.

### CODE REVIEW

File: src/main/java/gui/DiscoveryMainPanel.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+            switch (e.getActionCommand()) {
+                case "Preference":
+
+                    this.add(displayPage);
+                    break;
+                case "Location":
+                    this.add(displayPage);
+                    break;
+                case "Search":
+                   // SearchAsker asker = new SearchAsker();
```
Comment (sumiyuan) [inline src/main/java/gui/DiscoveryMainPanel.java:33]: can remove unused lines

File: src/main/java/gui/DiscoveryMainPanel.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    private SearchQuestionPanel panelS;
+
+
+    public DiscoveryMainPanel(int width, int height){
+        this.setBounds(0,0,width, height);
+        this.setLayout(new BoxLayout(this, BoxLayout.Y_AXIS));
+
+        DisplayPage displayPage = new DisplayPage(width,height);
+
+        panelS = new SearchQuestionPanel();
```
Comment (sumiyuan) [inline src/main/java/gui/DiscoveryMainPanel.java:18]: use a more appropriate variable name, like questionPanel or something else for easier reading

File: src/main/java/gui/OptionDisplayPanel.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        this.actionListeners = new ArrayList<>();
+        this.setBounds(0,0, width, height);
+        this.setLayout(new BoxLayout(this, BoxLayout.Y_AXIS));
+
+        JLabel label = new JLabel();
+        label.setText("Choose one from below");
+        label.setBounds((int) (width * 0.5), 0, 120, 20);
+        this.add(label);
+
+        Button btnP = new Button((int) (width*0.6), 125, 120, 60, "By Preference");
```
Comment (sumiyuan) [inline src/main/java/gui/OptionDisplayPanel.java:24]: use a more appropriate varaible name, like preferenceButton

File: src/main/java/gui/OptionDisplayPanel.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        btnP.setActionCommand("Preference");
+        btnP.addActionListener(e -> {
+            for(ActionListener listener: actionListeners){
+                listener.actionPerformed(e);
+            }
+        });
+        this.add(btnP);
+
+        Button btnL = new Button((int) (width*0.6), 195, 120, 60, "By Location");
```
Comment (sumiyuan) [inline src/main/java/gui/OptionDisplayPanel.java:33]: Some as above, use a moe appropriate variable name

File: src/main/java/gui/OptionDisplayPanel.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        btnL.setActionCommand("Location");
+        btnL.addActionListener(e -> {
+            for(ActionListener listener: actionListeners){
+                listener.actionPerformed(e);
+            }
+        });
+        this.add(btnL);
+
+        Button btnS = new Button((int) (width*0.6), 265, 120, 60, "By Search");
```
Comment (sumiyuan) [inline src/main/java/gui/OptionDisplayPanel.java:42]: same here

File: src/main/java/gui/SearchQuestionPanel.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+        this.setBounds(0, 0, width,height);
+
+        JLabel questionTitle = new JLabel();
+        questionTitle.setBounds(20, 20, 160, 20);
+        questionTitle.setText(viewModel.getIncomeQuestion());
+        this.add(questionTitle);
+
+        TextField textFieldLow = new TextField();
+        textFieldLow.setBounds(40, 80, 50, 20);
+        //textFieldLow.setBackground(Color.cyan);
```
Comment (sumiyuan) [inline src/main/java/gui/SearchQuestionPanel.java:31]: remove unused line

File: src/main/java/use_case_discovery/DiscoveryOptionChooser.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        private String marriageQuestion;
+        private String[] marriageOption;
+        private String hobbyQuestion;
+        private String[] hobbyOption;
+        private String relationshipQuestion;
+        private String[] relationshipOption;
+        private String petQuestion;
+        private String[] petOption;
+
+
```
Comment (sumiyuan) [inline src/main/java/use_case_discovery/DiscoveryOptionChooser.java:35]: try to group all setters and getters in seperate clusters, it makes the code easier to read and navigate

File: src/main/java/use_case_discovery/GenderFinder.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+import entity.User;
+
+import java.util.ArrayList;
+
+/**
+     * GenderFinder takes user's sexual orientation and filter other users based
+     * Documentary of possible sexual orientation:
+     *(possible sexual orientation): (return from GenderFinder)
+     * male: return only users identified themselves as male
```
Comment (sumiyuan) [inline src/main/java/use_case_discovery/GenderFinder.java:11]: should be more explicit in the return type in this instance it should say male : returns an ArrayList<User> of users that identifies as male

**Review by sumiyuan — COMMENTED** [review]
generally looks alright, many places needs more documentation to indicated what they do, and variable names should indicate what the variable is used for.

**Review by Caconoutt — APPROVED** [review]
The pull request is reviewed and approved by Camilla. The search discovery function is finalized, and the implementation aligns with the design pattern used in the preference discovery. The Javadoc added is clear to understand. Tests have been added and all successfully passed. Clean architecture has been followed in the code. Great job!

**Review by sumiyuan — APPROVED** [review]
Looks good to me, some of the files are lacking documentations but that can be added on later. Good unit tests, covers most of the functionality of the issue. Have to finalize that is going in the file and what is not ie removing all commented lines that are not needed.

**Review by Jennifer-xxx — APPROVED** [review]
This pull request is reviewed and approved by Jennifer Liu. The search discovery feature is well implemented based on clean architecture. The use of a database also shows the overall data persistence. The result of the program fulfills the expectation. The implementation of the feature combines well with the preference discovery feature. The tests are well-written and all pass. Overall looks good from my end!

**Pull request merged**